In [1]:
import pandas as pd
import math
import numpy as np

## Factorial

In [3]:
def factorial(x):
    if x == 1:
        return 1
    else:
        return x*factorial(x-1)

In [5]:
factorial(3)

6

In [6]:
factorial(0)

RecursionError: maximum recursion depth exceeded in comparison

In [7]:
math.factorial(0)

1

In [11]:
def true_factorial(x):
    
    def factorial(x):
        if x == 1:
            return 1
        else:
            return x*factorial(x-1)
    
    return int(factorial(x+1)/(x+1))
    

In [15]:
true_factorial(5)

120

n! is the number of ways we can arrange n objects

$$ x! = x \times (x-1) \times (x-2) \times ... \times 2 \times 1 $$ 

$$ x! = \frac{(x+1)!}{(x+1)} = \frac{(x+1) \times x \times (x-1) \times (x-2) \times ... \times 2 \times 1}{x+1} $$ \
$$ 4! = \frac{5!}{5} = \frac{120}{5} = 24 $$ \
$$ 3! = \frac{4!}{4} = \frac{24}{4} = 6 $$ \
$$ 2! = \frac{3!}{3}  = \frac{6}{3} = 2  $$ \
$$ 1! = \frac{2!}{2} = \frac{2}{2} = 1 $$ \
$$ 0! = \frac{1!}{1} = \frac{1}{1} = 1 $$

## Fibonacci numbers

The Fibonacci numbers, commonly denoted F(n) form a sequence, called the Fibonacci sequence, such that each number is the sum of the two preceding ones, starting from 0 and 1.

$$ F(0) = 0,   \ $$
$$ F(1) = 1 $$ \
$$ F(N) = F(N - 1) + F(N - 2), for N > 1 $$


In [25]:
def fibonacci(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)

In [26]:
list(fibonacci(i) for i in range(10))

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [15]:
def fib(n):
    if n in range(2):
        return n
    else:
        return fib(n-1) + fib(n-2)

In [16]:
for i in range(10):
    print(fib(i))

0
1
1
2
3
5
8
13
21
34


In [19]:
list(fib(i) for i in range(10))

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]

In [18]:
fibonacci(5)

5

In [22]:
def pisano_period(d, length):
    fib_sequence = list(fib(i) for i in range(length))
    remainders = [num%d for num in fib_sequence]
    
    return remainders

In [23]:
pisano_period(5, 20)

[0, 1, 1, 2, 3, 0, 3, 3, 1, 4, 0, 4, 4, 3, 2, 0, 2, 2, 4, 1]